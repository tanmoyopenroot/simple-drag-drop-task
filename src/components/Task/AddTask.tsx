import * as React from 'react';
import { IoIosAdd } from 'react-icons/io';

import Button from '../Button';
import { AddTaskWrapper } from './Task.styles';

interface IAddActionsProps {
  onAddClick: () => void;
}

const AddTaskActions: React.FC<IAddActionsProps> = (props) => {
  const { onAddClick } = props;

  return (
    <AddTaskWrapper>
      <Button action={onAddClick}>
        <React.Fragment>
          <IoIosAdd />
          Add Task
        </React.Fragment>
      </Button>
    </AddTaskWrapper>
  );
};

export default React.memo(AddTaskActions);
