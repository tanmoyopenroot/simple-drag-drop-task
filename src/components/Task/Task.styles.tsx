import styledComponents from 'styled-components';

const GROUPED_COLUMN = styledComponents.div`
  display: flex;
  flex-direction: column;
`;

const GROUPED_ROW = styledComponents.div`
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  align-items: center;
`;

export const TaskWrapper = styledComponents(GROUPED_ROW)`
  background: ${props => props.theme.SECONDARY_BACKGROUND_IMAGE};
  border-radius: 8px;
  cursor: grab;
  padding: .66rem 1rem;
  position: relative;
`;

export const TaskActionWrapper = styledComponents(GROUPED_COLUMN)`
  color: ${props => props.theme.PRIMARY_TEXT};
  cursor: pointer;
`;

export const TaskBody = styledComponents.div`
  color: ${props => props.theme.PRIMARY_TEXT};
  font-size: ${props => props.theme.FONT.MEDIUM};
`;

export const AddTaskWrapper = styledComponents(GROUPED_ROW)`
  background-image: ${props => props.theme.ACTIONABLE_BACKGROUND_IMAGE};
  border-radius: 8px;
  cursor: pointer;
  padding: .66rem 1rem;
  position: relative;
  margin-top: .5rem;

  button {
    color: ${props => props.theme.PRIMARY_TEXT};
  };
`;
