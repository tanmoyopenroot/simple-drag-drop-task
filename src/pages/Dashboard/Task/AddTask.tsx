import * as React from 'react';
import { IoIosAdd } from 'react-icons/io';

import Button from '../../../components/Button';
import TextHeader from '../../../components/TextHeader';
import Input from '../../../components/Input';
import Dialog from '../../../components/Dialog';
import { GroupedRow } from '../../../common/styles';
import {
  InputWrapper,
} from './Task.styles';

interface IAddActionsProps {
  onCreate: (title: string) => void;
}

interface IAddActionState {
  body: string;
}

class AddTaskActions extends React.Component<IAddActionsProps, IAddActionState> {
  private dialogRef = React.createRef<HTMLDialogElement>();

  public state: IAddActionState = {
    body: '',
  };

  handleOnAddPanelClick = () => {
    if (this.dialogRef.current) {
      this.dialogRef.current.showModal();
    }
  }

  handleCloseDialog = () => {
    if (this.dialogRef.current) {
      this.dialogRef.current.close();
    }
  }

  handleOnCreate = () => {
    const { body } = this.state;
    const { onCreate } = this.props;

    if (body) {
      onCreate(body);
      this.handleCloseDialog();
    }
  }

  handleOnChange = (value: string) => {
    this.setState({ body: value });
  }

  public render() {
    const { body } = this.state;

    return (
      <React.Fragment>
        <Button onClick={this.handleOnAddPanelClick}>
          <React.Fragment>
            <IoIosAdd />
            Add Task
          </React.Fragment>
        </Button>
        <Dialog ref={this.dialogRef}>
          <React.Fragment>
            <TextHeader>
              Enter task Info
            </TextHeader>
            <InputWrapper>
              <Input
                autoFocus={true}
                defaultValue={body}
                placeholder="Add Task"
                onChange={this.handleOnChange}
              />
            </InputWrapper>
            <GroupedRow justifyContent="space-evenly">
              <Button
                width="200px"
                onClick={this.handleCloseDialog}
              >
                Close Dialog
              </Button>
              <Button
                width="200px"
                onClick={this.handleOnCreate}
              >
                Create Task
              </Button>
            </GroupedRow>
          </React.Fragment>
        </Dialog>
      </React.Fragment>
    );
  }
}

export default React.memo(AddTaskActions);
