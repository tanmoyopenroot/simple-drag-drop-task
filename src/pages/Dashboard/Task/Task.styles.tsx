import styledComponents from 'styled-components';
import {
  GroupedRow,
  GroupedColumn,
} from '../../../common/styles';

export const TaskWrapper = styledComponents(GroupedRow)`
  background: ${props => props.theme.SECONDARY_BACKGROUND_IMAGE};
  border-radius: ${props => props.theme.RADIUS};
  cursor: grab;
  padding: .66rem 1rem;
  position: relative;
`;

export const TaskActionWrapper = styledComponents(GroupedColumn)`
  color: ${props => props.theme.PRIMARY_TEXT};
  cursor: pointer;
`;

export const TaskBody = styledComponents.div`
  color: ${props => props.theme.PRIMARY_TEXT};
  font-size: ${props => props.theme.FONT && props.theme.FONT.MEDIUM};
`;

export const InputWrapper = styledComponents.div`
  background: ${props => props.theme.SECONDARY_BACKGROUND_IMAGE};
  border-radius: ${props => props.theme.RADIUS};
  cursor: grab;
  padding: .66rem 1rem;
  position: relative;
  margin: 20px 0px;
`;

export const AddTaskWrapper = styledComponents.div`
  margin-top: 20px;
`;
