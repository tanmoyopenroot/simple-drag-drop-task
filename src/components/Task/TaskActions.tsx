import * as React from 'react';
import { IoMdCreate } from 'react-icons/io';

import { TaskActionWrapper } from './Task.styles';

interface IItemActionsProps {
  onEditClick: () => void;
}

const ItemActions: React.FC<IItemActionsProps> = (props) => {
  const { onEditClick } = props;

  return (
    <TaskActionWrapper>
      <IoMdCreate onClick={onEditClick} />
    </TaskActionWrapper>
  );
};

export default React.memo(ItemActions);
