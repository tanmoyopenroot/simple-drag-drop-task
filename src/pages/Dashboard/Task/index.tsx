import Task from './Task';
import AddTask from './AddTask';

export {
  Task,
  AddTask,
};
