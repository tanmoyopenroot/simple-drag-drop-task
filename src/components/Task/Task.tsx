import * as React from 'react';

import Input from '../Input';
import TaskActions from './TaskActions';
import {
  TaskBody,
  TaskWrapper,
} from './Task.styles';

export interface ITaskProps {
  id: string;
  body?: string;
}

export interface ITaskState {
  isEditing: boolean;
}

class Task extends React.Component<ITaskProps, ITaskState> {
  public static defaultProps: Omit<ITaskProps, 'id'> = {
    body: 'Enter new task...',
  };

  public state: ITaskState = {
    isEditing: true,
  };

  public shouldComponentUpdate(nextProps: ITaskProps, nextState: ITaskState) {
    const {
      id,
      body,
    } = this.props;
    const { isEditing } = this.state;

    return id !== nextProps.id ||
      body !== nextProps.body ||
      isEditing !== nextState.isEditing;
  }

  private toggleEdit = () => {
    const { isEditing } = this.state;

    this.setState({
      isEditing: !isEditing,
    });
  }

  private handleOnBlur = (value: string) => {
    this.toggleEdit();
  }

  private getInput = () => {
    const { body } = this.props;
    const { isEditing } = this.state;

    return isEditing ? (
      <Input
        autoFocus={true}
        defaultValue={body}
        onBlur={this.handleOnBlur}
      />
    ) : (
      <TaskBody>
        {body}
      </TaskBody>
    );
  }

  render() {
    const { id } = this.props;

    return(
      <TaskWrapper id={id}>
        {this.getInput()}
        <TaskActions onEditClick={this.toggleEdit} />
      </TaskWrapper>
    );
  }
}

export default Task;
