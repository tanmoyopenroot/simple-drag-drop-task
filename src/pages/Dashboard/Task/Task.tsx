import * as React from 'react';

import Input from '../../../components/Input';
import TaskActions from './TaskActions';
import {
  TaskBody,
  TaskWrapper,
} from './Task.styles';

export interface ITaskProps {
  id: string;
  body?: string;
  onChange: (body: string, id: string) => void;
  onDragStart: (event: React.MouseEvent<HTMLDivElement>, id: string) => void;
  onDragEnd: (event: React.MouseEvent<HTMLDivElement>, id: string) => void;
}

export interface ITaskState {
  isEditing: boolean;
}

export interface IDefaultProps {
  body?: string;
}

class Task extends React.Component<ITaskProps, ITaskState> {
  public static defaultProps: IDefaultProps = {
    body: '',
  };

  public state: ITaskState = {
    isEditing: false,
  };

  public shouldComponentUpdate(nextProps: ITaskProps, nextState: ITaskState) {
    const {
      id,
      body,
    } = this.props;
    const { isEditing } = this.state;

    return id !== nextProps.id ||
      body !== nextProps.body ||
      isEditing !== nextState.isEditing;
  }

  private toggleEdit = () => {
    const { isEditing } = this.state;

    this.setState({
      isEditing: !isEditing,
    });
  }

  private handleOnBlur = (value: string) => {
    const {
      id,
      onChange,
    } = this.props;

    onChange(value, id);
    this.toggleEdit();
  }

  private handleDragStart = (event: React.MouseEvent<HTMLDivElement>) => {
    const {
      id,
      onDragStart,
    } = this.props;

    onDragStart(event, id);
  }

  private handleDragEnd = (event: React.MouseEvent<HTMLDivElement>) => {
    const {
      id,
      onDragEnd,
    } = this.props;

    onDragEnd(event, id);
  }

  private getInput = () => {
    const { body } = this.props;
    const { isEditing } = this.state;

    return isEditing ? (
      <Input
        autoFocus={true}
        defaultValue={body}
        onBlur={this.handleOnBlur}
      />
    ) : (
      <TaskBody>
        {body}
      </TaskBody>
    );
  }

  render() {
    const { id } = this.props;

    return(
      <TaskWrapper
        id={id}
        draggable={true}
        onDragStart={this.handleDragStart}
        onDragEnd={this.handleDragEnd}
      >
        {this.getInput()}
        <TaskActions onEditClick={this.toggleEdit} />
      </TaskWrapper>
    );
  }
}

export default Task;
