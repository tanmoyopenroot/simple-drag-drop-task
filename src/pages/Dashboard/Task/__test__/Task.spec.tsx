import * as React from 'react';
import { mount } from 'enzyme';
import initalState from '../../../../data/inital-state';

import TaskActions from '../TaskActions';
import Task, {
  ITaskProps,
  ITaskState,
} from '../Task';
import {
  TaskBody,
  TaskWrapper,
} from '../Task.styles';

describe('Test <Task />', () => {
  const panelId = initalState.panels.panelsID[0];
  const panel = initalState.panels.panelsHash[panelId];
  const taskId = initalState.tasks.tasksIDByPanel[panel.id][0];
  const task = initalState.tasks.tasksHash[taskId];

  const props: ITaskProps = {
    id: task.id,
    body: task.body,
    onDrag: jest.fn(),
    onChange: jest.fn(),
    onDragStart: jest.fn(),
  };
  const wrapper = mount<Task, ITaskState, ITaskProps>(
    <Task {...props}/>,
  );
  const wrapperInstance = wrapper.instance();

  it('should be able to render with props', () => {
    expect(wrapperInstance.props.id).toEqual(props.id);
    expect(wrapperInstance.props.body).toEqual(props.body);
    expect(wrapperInstance.props.onDrag).toEqual(props.onDrag);
    expect(wrapperInstance.props.onChange).toEqual(props.onChange);
    expect(wrapperInstance.props.onDragStart).toEqual(props.onDragStart);

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should render task', () => {
    expect(wrapperInstance.state.isEditing).toBeFalsy();
    expect(wrapper.find(TaskWrapper).exists()).toBeTruthy();

    expect(
      wrapper
        .find(TaskWrapper)
        .find(TaskActions)
        .exists(),
    ).toBeTruthy();

    expect(
      wrapper
        .find(TaskWrapper)
        .find(TaskBody)
        .exists(),
    ).toBeTruthy();

    expect(
      wrapper
      .find(TaskWrapper)
      .find(TaskBody),
    )
    .toHaveLength(1);

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should call toggleEdit method', () => {
    const wrapperInstance = wrapper.instance() as Task;

    expect(wrapperInstance.state.isEditing).toBeFalsy();
    wrapperInstance.toggleEdit();
    expect(wrapperInstance.state.isEditing).toBeTruthy();

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should call handleOnBlur method', () => {
    const { id } = props;

    const wrapperInstance = wrapper.instance() as Task;

    wrapperInstance.handleOnBlur('Edited Task');
    expect(props.onChange).toBeCalledWith('Edited Task', id);
    expect(props.onChange).toBeCalledTimes(1);

    expect(wrapper.html()).toMatchSnapshot();
  });
});
