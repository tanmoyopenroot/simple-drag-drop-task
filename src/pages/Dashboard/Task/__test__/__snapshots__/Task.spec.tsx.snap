// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test <Task /> should be able to render with props 1`] = `"<div id=\\"_qvf8cg4dn\\" draggable=\\"true\\" class=\\"sc-bdVaJa sc-htpNat iCGEAK\\"><div class=\\"sc-ifAKCX gJLhNS\\">Body Of Task A</div><div class=\\"sc-bwzfXH sc-bxivhb eyLEbJ\\"><svg stroke=\\"currentColor\\" fill=\\"currentColor\\" stroke-width=\\"0\\" viewBox=\\"0 0 512 512\\" height=\\"1em\\" width=\\"1em\\" xmlns=\\"http://www.w3.org/2000/svg\\"><path d=\\"M64 368v80h80l235.727-235.729-79.999-79.998L64 368zm377.602-217.602c8.531-8.531 8.531-21.334 0-29.865l-50.135-50.135c-8.531-8.531-21.334-8.531-29.865 0l-39.468 39.469 79.999 79.998 39.469-39.467z\\"></path></svg></div></div>"`;

exports[`Test <Task /> should call handleOnBlur method 1`] = `"<div id=\\"_qvf8cg4dn\\" draggable=\\"true\\" class=\\"sc-bdVaJa sc-htpNat iCGEAK\\"><div class=\\"sc-ifAKCX gJLhNS\\">Body Of Task A</div><div class=\\"sc-bwzfXH sc-bxivhb eyLEbJ\\"><svg stroke=\\"currentColor\\" fill=\\"currentColor\\" stroke-width=\\"0\\" viewBox=\\"0 0 512 512\\" height=\\"1em\\" width=\\"1em\\" xmlns=\\"http://www.w3.org/2000/svg\\"><path d=\\"M64 368v80h80l235.727-235.729-79.999-79.998L64 368zm377.602-217.602c8.531-8.531 8.531-21.334 0-29.865l-50.135-50.135c-8.531-8.531-21.334-8.531-29.865 0l-39.468 39.469 79.999 79.998 39.469-39.467z\\"></path></svg></div></div>"`;

exports[`Test <Task /> should call toggleEdit method 1`] = `"<div id=\\"_qvf8cg4dn\\" draggable=\\"true\\" class=\\"sc-bdVaJa sc-htpNat iCGEAK\\"><input type=\\"text\\" placeholder=\\"\\" class=\\"sc-gzVnrw goKrNz\\" value=\\"Body Of Task A\\"><div class=\\"sc-bwzfXH sc-bxivhb eyLEbJ\\"><svg stroke=\\"currentColor\\" fill=\\"currentColor\\" stroke-width=\\"0\\" viewBox=\\"0 0 512 512\\" height=\\"1em\\" width=\\"1em\\" xmlns=\\"http://www.w3.org/2000/svg\\"><path d=\\"M64 368v80h80l235.727-235.729-79.999-79.998L64 368zm377.602-217.602c8.531-8.531 8.531-21.334 0-29.865l-50.135-50.135c-8.531-8.531-21.334-8.531-29.865 0l-39.468 39.469 79.999 79.998 39.469-39.467z\\"></path></svg></div></div>"`;

exports[`Test <Task /> should render task 1`] = `"<div id=\\"_qvf8cg4dn\\" draggable=\\"true\\" class=\\"sc-bdVaJa sc-htpNat iCGEAK\\"><div class=\\"sc-ifAKCX gJLhNS\\">Body Of Task A</div><div class=\\"sc-bwzfXH sc-bxivhb eyLEbJ\\"><svg stroke=\\"currentColor\\" fill=\\"currentColor\\" stroke-width=\\"0\\" viewBox=\\"0 0 512 512\\" height=\\"1em\\" width=\\"1em\\" xmlns=\\"http://www.w3.org/2000/svg\\"><path d=\\"M64 368v80h80l235.727-235.729-79.999-79.998L64 368zm377.602-217.602c8.531-8.531 8.531-21.334 0-29.865l-50.135-50.135c-8.531-8.531-21.334-8.531-29.865 0l-39.468 39.469 79.999 79.998 39.469-39.467z\\"></path></svg></div></div>"`;
